### PR TITLE
fix: allow passing a filterset class to the filterset factory

### PR DESCRIPTION
**Pull request by b1rger**
I've tried to reuse the variable name terminology from `django.forms.models.modelform_factory`

Closes: #1631

### CODE REVIEW

**Review by carltongibson — COMMENTED** [review]
Hi @b1rger — this looks good. 

Could you add a test or two using the new argument? (Specifically, interested in the base filter set defining some properties to be reused, fields and meta options.)

**Comment by b1rger** [comment]
> Could you add a test or two using the new argument? Specifically, interested in the base filter set defining some properties to be reused, fields and meta options: inheritance of meta options needs testing. Also we should doc this.

done

**Comment by b1rger** [comment]
FTR: currently the `Meta.fields` attribute of the base filterset is always overwritten by the `fields` that is passed to the factory (which defaults to `ALL_FIELDS`) - that means if you want to provide use a custom filterset class together with custom field definitions, you would have to both provide a custom filtersetclass *and* pass the fields to the factory. It might be more usable if the `fields` argument of the factory got only used if the base filterset Meta class does not have one. Or if the `fields` argument of the factory is None, but that would break the method signature... maybe only set fields `if fields is None and not hasattr(filterset.Meta, "fields")`...

**Comment by carltongibson** [comment]
It's probably worth comparing how `modelform_factory` does it.

**Comment by b1rger** [comment]
> It's probably worth comparing how `modelform_factory` does it.

They are passing `fields=None` to the factory and then:
```
    attrs = {"model": model}
    if fields is not None:
        attrs["fields"] = fields
```
I can update the code accordingly, if its ok to change the `fields` argument default from `ALL_FIELDS` to `None`

**Comment by carltongibson** [comment]
They're subclassing the `Meta` class, so that defined options are inherited.

**Comment by b1rger** [comment]
ack, the `Meta.fields` in the passed `filterset` now trumps the `fields`

**Comment by b1rger** [comment]
> ack, the `Meta.fields` in the passed `filterset` now trumps the `fields`

Hm, scratch that, i pushed too fast, i'll have to rethink the logic...

**Comment by b1rger** [comment]
Oke, I think I found an approach that should not lead to surprises: if the `fields` argument is not `None` it overrides everything. If it *is* `None`, the `Meta.fields` of the filterset argument is used, if it exists and is not `None`. If thats also not the case, `fields` is set to `ALL_FIELDS`

File: tests/test_filterset.py [hunk 1]
```diff
@@ -715,6 +711,73 @@ class F(A, B):
...
+                    },
+                }
+
+        filterset = filterset_factory(Article, FilterSetBase)
+
+        f = Article._meta.get_field("author")
+        result, params = filterset.filter_for_lookup(f, "isnull")
+        self.assertEqual(result, BooleanFilter)
+
+    def test_filtesret_factory_base_filter_meta_inheritance_exclude(self):
```
Comment (carltongibson) [inline tests/test_filterset.py:773]: Could you check the typos in the test method names please? (Same with test above too.)

**Review by carltongibson — COMMENTED** [review]
OK, I think this looks good. Just the one comment.

**Pull request closed without merging**